### DISCUSSION TITLE
.with and polymorphic relationships

### DISCUSSION THREAD

**Discussion by tintin10q** [CodeDredd/pinia-orm#1606] [Ideas]
Hi there,

I have now arrived at the point in my pinia orm journey that I need to start using polymorphic relationships. Things so far have been good. Besides the `.with`.

I have a schema like this:

```ts
class HumanSavedMovie extends Model {
  static entity = "HumanSavedEpisode";
  static primaryKey = ["human_id", "movie_id"];

  static fields() {
    return {
      movie_id: this.uid(),
      human_id: this.uid(),
      added_at: this.attr(utcsec()),
      human: this.belongsTo(Human, "human_id"),
      movie: this.belongsTo(Movie, "episode_id"),
      saved: this.morphOne(Saved, "id", "T", "movie_id").onDelete("cascade"),
    };
  }
 get saved_id() {
    return this.movie_id;
  }
}

class HumanSavedEpisode extends Model {
  static entity = "HumanSavedEpisode";
  static primaryKey = ["human_id", "episode_id"];

  static fields() {
    return {
      episode_id: this.uid(),
      human_id: this.uid(),
      added_at: this.attr(utcsec()),
      human: this.belongsTo(Spotify_Human, "human_id"),
      episode: this.belongsTo(Spotify_Episode, "episode_id"),
      saved: this.morphOne(Saved, "id", "T", "episode_id").onDelete("cascade"),
    };
  }
 get saved_id() {
    return this.episode_id;
  }
}

// and now the Saved:

class Saved<Item extends SavebleT> extends Model {
  static entity = "saved";

  static fields() {
    return {
      // this is like item id
      id: this.uid(),
      added_at: this.number(utcsec),
      T: this.string(""),
      item: this.morphTo(
        [HumanSavedMovie, HumanSavedEpisode],
        "id",
        "T",
        "saved_id"
      ),
    };

  declare id: string;
  /* type */
  declare T: SavebleT;
  declare item: Item extends "HumanSavedMovie"
    ? HumanSavedMovie
    : Item extends "HumanSavedEpisode"
    ? HumanSavedEpisode"
    : never;
}
```

The first issue I encountered is that all my HumanSavedX models had different primary keys. This does not work with polymorphic relationships because you can only specify one custom ownerKey. I could work around this by specifying the `saved_id` getter on  `HumanSavedMovie` and `HumanSavedEpisode`. Maybe being able to pass a list of ownerKeys the same length as the first input to MorphTo is a good idea.

However, this trick does not work for relationships. This query does not work:

```ts
const saved = useRepo(Saved);
const items = saved.query().with("item", (item) => item.with("movie")).get()
```

This one gives the error that the `movie` relationship does not exist on `HumanSavedEpisode`.

and neither this query works:

```ts
const saved = useRepo(Saved);
const items = saved.query().with("item", (item) => item.with("episode")).get()
```

This one gives the error that the `episode` relationship does not exist on `HumanSavedMovie`.

-----

One solution of course would be to rename movie and episode to `item`. 
I could still use the composite primary key but it would be less clear. Also, this just moves the problem. My `movie` model happens to have an `artists` relationship. This relationship does not exist on episodes. **This means I can not query the artists on the movie!**

This error means that, yes you can link multiple entities to the same entity with polymorphic relationships. The relationships of the entities that you link to have to be quite similar for it to be useful. 

Could we think about a way to not throw an error with polymorphic relationships and queries like this?

```ts
const items = saved.query().with("item", (item) => item.with("episode").with("movie")).get()
```

I would like to have a discussion about this.

**Comment by tintin10q** [comment]
One possible direction could be something like this:

```ts
const all_saved = computed(() =>
  saved
    .query()
    .with("saved_item", (saved_item) =>
      saved_item
        .whereHas("movie", (query) => query.with("artists"))
        .orWhereHas("episode", (query) => query.with("episode"))
    )
    .get()
);
```

Where you specify the relationships based on having certain other relationships.

**Comment by tintin10q** [comment]
Another solution could be to make fake relationships on other models. Like giving my episodes an artist relationship which will never have anything. But this does not seem very scalable.

**Comment by tintin10q** [comment]
Just pining so you see this. @CodeDredd

**Reply by CodeDredd** [reply]
Soon as i have time i look onto it.

**Reply by tintin10q** [reply]
Actually I think https://pinia-orm.codedredd.de/guide/model/single-table-inheritance is exactly what I need.
